ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx

```diff
@@ -1,26 +1,152 @@
+  const [thumbnailState, setThumbnailState] = useState<Blob[]>([]);
+
+  const handleFileDeletion = (fileIndex: number) => {
+    setFormState(prevState => {
+      const filesClone = [...files];
+      filesClone.splice(fileIndex, 1);
+      const updatedState = {
+        [name]: {
+          ...prevState[name],
+          value: [...filesClone],
+        },
+      };
+
+      return { ...prevState, ...updatedState } as Record<string, FieldData>;
+    });
+
+    setThumbnailState(prevState => {
+      const thumbnailClone = [...prevState];
+      thumbnailClone.splice(fileIndex, 1);
+
+      return thumbnailClone;
+    });
+  };
+
+  useEffect(() => {
+    Promise.all(
+      files.map(async file => {
+        return new Promise(resolve => {
+          const fileReader = new FileReader();
+          fileReader.onloadend = () => {
+            resolve(fileReader.result);
+          };
+          fileReader.readAsDataURL(file);
+        });
+      }),
+    ).then(result => setThumbnailState(result as SetStateAction<Blob[]>));
+  }, [files]);
+
+  const listItems = files.map((file: File, index: number) => {
+    const key = `${index}-${file.name}`;
+    const thumbnailSrc = thumbnailState[index];
+
+    return (
+      <li key={key}>
+        <div>{file.name}</div>
+        <button type="button" onClick={() => handleFileDeletion(index)}>
+          {navigationIcons.cross}
+        </button>
+        <img src={`${thumbnailSrc}`} alt="" />
+      </li>
+    );
+  });
+  return <ul>{listItems}</ul>;
+};
+
+export default ({ id, name, inputState, describedBy }: InputProps) => {
   const { isValid, required } = inputState;
-  return (
-    <input
-      id={id}
-      name={name}
-      type="file"
-      onChange={e =>
-        e.target.files && handleChange(e.target.name, e.target.files)
+  const { setFormState } = useFormContext();
+  const inputRef = useRef<HTMLInputElement>(null);
+  const files = inputState.value;
+
+  const handleFileChange = (event: ChangeEvent<HTMLInputElement>) => {
+    setFormState(prevState => {
+      const value = event.target.files;
+      if (!value) return { ...prevState };
+
+      const fileListLength = value.length;
+      const filesArray = [];
+
+      const isInitialUpload = prevState[name].value === '';
+
+      for (let fileIndex = 0; fileIndex < fileListLength; fileIndex += 1) {
+        const file = value?.item(fileIndex);
+        if (file) filesArray.push(file);
       }
-      multiple
-      aria-invalid={!isValid}
-      aria-required={required}
-      aria-describedby={describedBy}
-    />
+
+      const updatedState = {
+        [name]: {
+          ...prevState[name],
+          ...(isInitialUpload && {
+            value: [...filesArray],
+          }),
+          ...(!isInitialUpload && {
+            value: [...(prevState[name].value as File[]), ...filesArray],
+          }),
+        },
+      };
+
+      return { ...prevState, ...updatedState };
+    });
+  };
```

Comment:
Took a bit of inspiration from https://github.com/KunalN25/multiple-file-uploads/blob/main/src/App.js and came up with this to make this function a bit simpler:
```
  const { isValid, required } = inputState;
  const { setFormState } = useFormContext();
  const inputRef = useRef<HTMLInputElement>(null);
  const filesInState = inputState.value as File[];

  const handleFileChange = (e: ChangeEvent<HTMLInputElement>) => {
    const chosenFiles = Array.prototype.slice.call(e.target.files) as File[];
    const uploaded = [...filesInState];

    chosenFiles.some(file => {
      if (uploaded.findIndex(f => f.name === file.name) === -1) {
        uploaded.push(file);
      }
      return null;
    });

    setFormState(prevState => {
      return {
        ...prevState,
        [name]: { ...prevState[name], value: uploaded },
      };
    });
  };
```

Seems to work when I try it, but I'm not sure if it covers all the edge cases!

Reply:
In fact you may just be able to use `handleChange` thats passed in the component args with this change:

```
  const handleFileChange = (e: ChangeEvent<HTMLInputElement>) => {
    const chosenFiles = Array.prototype.slice.call(e.target.files) as File[];
    const uploaded = [...filesInState];

    chosenFiles.some(file => {
      if (uploaded.findIndex(f => f.name === file.name) === -1) {
        uploaded.push(file);
      }
      return null;
    });

    handleChange(name, uploaded);
  };
```

Reply:
I haven't tried actually submitting these files, so this all may be non-applicable 😅

Reply:
Yeah was on the fence with `Array.prototype.slice.call` partly because I thought being explicit would help understand why can't just manipulate the FileList, happy to change to it

Reply:
Yea I needed to read up on it myself. I think overall its maybe a bit simpler to reason about the function, but up to other reviewers as well if the prototype function is a bit too magic-y.

Reply:
I've opted to change it, and just added a comment above it

Reply:
Hmm just thought about that some method, there might be a case where a user is uploading a file with the same name, but are different? And I don't think we have an error state for that, as the old uploader allowed uploading of multiple of the same file

Reply:
Does the `file.name` include the extension as well? If it does, I think it shouldn't allow that scenario, don't think you can have 2 files named the same with the same extension.

Reply:
You can if it lives in a different directory, as the file object contains a reference to the file location on a users machine. So if bob wanted to upload `/downloads/img.jpg`, then opens the uploader again and adds `/desktop/img.jpg` it would let them do it

Reply:
So i guess we can use that value

Reply:
Ah yea right enough. Tbh I think its fine to let them upload multiple of the same file. Probably an edge case that we don't need to worry too much about.